### PR TITLE
hubble: Add Support for filtering on HTTP headers

### PR DESCRIPTION
**Pull request by ChrsMark**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Implements the server side changes for  https://github.com/cilium/hubble/issues/776

```release-note
hubble: Add Support for filtering on HTTP headers
```

### How to test this manually

1. Deploy cilium from source with `make kind-image && make kind-install-cilium` (note the `hubble-export-file-path` should be set for some weird reason that is not super clear to me, this is how the code flow works at https://github.com/cilium/cilium/blob/f078737ed9bc0ad38cee0f6df5dd9624d83f4aad/daemon/cmd/hubble.go#L173)
2. Enable hubble with `cilium hubble enable --chart-directory=/home/chrismark/go/src/github.com/cilium/cilium/install/kubernetes/cilium`
3. Apply a L7 example endpoint with `kubectl apply -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/sw_l3_l4_l7_policy.yaml`
4. Port forward hubble relay svc with `kubectl port-forward -n kube-system svc/hubble-relay --address 0.0.0.0 --address :: 4245:80`
5. Deploy the sample app with `kubectl create -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/http-sw-app.yaml`. From https://docs.cilium.io/en/stable/gettingstarted/demo/#starwars-demo

-> Build hubble CLI from https://github.com/cilium/hubble/pull/1277 after replacing the `cilium` dependency in the `go.mod` file with `replace github.com/cilium/cilium => /home/chrismark/go/src/github.com/cilium/cilium`.

7. Observe the flows with `./hubble observe --pod tiefighter --http-header Foo:bar -f`
8. Execute a *matching* curl request with `kubectl exec tiefighter -- curl -H "Foo: bar" -s -XPOST "http://deathstar.default.svc.cluster.local/v1/request-landing"`
9. Verify that the flow is observed:
```console
Oct 29 19:46:06.337: default/tiefighter:39348 (ID:49781) -> default/deathstar-7848d6c4d5-9z2wr:80 (ID:43688) http-request FORWARDED (HTTP/1.1 POST http://deathstar.default.svc.cluster.local/v1/request-landing)
```
10. Execute a *non-matching* curl request with `kubectl exec tiefighter -- curl -H "Foo: baz" -s -XPOST "http://deathstar.default.svc.cluster.local/v1/request-landing"`
11. Verify that the flow is not observed.

### CODE REVIEW

**Review by gandro — APPROVED** [review]
Looks good to me, thanks!

File: pkg/hubble/filters/http.go [hunk 1]
```diff
@@ -210,5 +231,18 @@ func (h *HTTPFilter) OnBuildFilter(ctx context.Context, ff *flowpb.FlowFilter) (
...
 
+	if ff.GetHttpHeader() != nil {
+		if !httpMatchCompatibleEventFilter(ff.GetEventType()) {
+			return nil, errors.New("filtering by http headers requires " +
+				"the event type filter to only match 'l7' events")
+		}
+
+		headerf, err := filterByHTTPHeaders(ff.GetHttpHeader())
+		if err != nil {
+			return nil, fmt.Errorf("invalid http method filter: %v", err)
```
Comment (ghost) [inline pkg/hubble/filters/http.go]:
```suggestion
			return nil, fmt.Errorf("invalid http header filter: %v", err)
```

**Review by ghost — CHANGES_REQUESTED** [review]
Thank you for tackling this @ChrsMark. The error msg needs to be updated, but otherwise looks good to me.

File: pkg/hubble/filters/http.go (hunk 1 above)
Comment (ChrsMark) [reply pkg/hubble/filters/http.go]: Ah good catch, thanks :)!

**Review by ghost — APPROVED** [review]
lgtm

**Comment by ChrsMark** [comment]
Hey @lambdanis @gandro ! Anything else missing here?

**Comment by gandro** [comment]
Required pipeline [ci-runtime](https://github.com/cilium/cilium/actions/runs/6695681611/job/18256929520) seems to have timed out. I'm going to restart it.

**Comment by squeed** [comment]
@ChrsMark the failing job has has its timeout bumped. I'm rebasing your PR and hopefully it all passes :-)

**Pull request merged**